### ISSUE TITLE
A few years after #701 and PyTorch is still using implicit __all__ imports.

### ISSUE THREAD

**Issue by cardoso-neto** [pytorch/pytorch#49416]
> i dont think we'll be fixing this. (i dont know if there's a way to do it either).

_Originally posted by @soumith in https://github.com/pytorch/pytorch/issues/701#issuecomment-329026903_

> > Maybe one could change `from torch._C import *` to `from torch._C import from_numpy, ...`?
> This seems simple enough. Are we certain this is the way to go? 
> I thought we'd need type hints (which would be way harder to convince people to use).
> Should we open another issue to discuss @moi90's solution?

_Originally posted by @cardoso-neto in https://github.com/pytorch/pytorch/issues/701#issuecomment-485272641_

A quick search for `import *` yields hundreds results.

If requested by a team member, I'd gladly do the tedious work replacing the import statements and pull requesting. :)
I would probably just need some pointers on how to do find out what is being imported from which C module, but I'd make quick work of it.
Long import statements could be indented very cleanly following [psf/black's](https://github.com/psf/black) style guides.

**Comment by ezyang** [comment]
I think since #701 we have more tools in our toolbox to fix this problem. For `torch._C` imports, it is feasible for us to code generate the list of identifiers that should be imported in this situations. In other situations where the import list is for some fixed set of identifiers that were not codegenned, I don't have problems with replacing them with explicit imports.

There is also a reverse problem, where some modules do `from m import f` and accidentally expose f in their context.

**Comment by rgommers** [comment]
https://github.com/asmeurer/removestar may also be useful - it worked well in cleaning `import *` from the Numba code base.

**Comment by cardoso-neto** [comment]
@rgommers Does Numba use C code as well?
I've come across a few other tools that also promise to do that, but I cannot speak to their effectiveness.
[github/autoflake/autostar.py](https://github.com/fake-name/autoflake/blob/master/autostar.py)
[github/quentinsf/dewildcard](https://github.com/quentinsf/dewildcard)

> modules do from m import f and accidentally expose f in their context

@ezyang Well, great. This is bad and we would avoid it by using explicit imports.

We could also enable a check on the pull requests so it'll flag bad star imports. 
The PyLint  config `W0614` checks for _Unused import from wildcard_ and the `W0401`  checks for _Wildcard import_.
Source: http://pylint-messages.wikidot.com/all-codes

**Comment by ezyang** [comment]
Yes. So I'm OK with helping review some exploratory work on this. The "abort" button I want you to press is if you have to manually type in a ton of autogenerated stuff; let's nail some of the easier to fix sites first.

Do note that most of FB will be going on Winter holidays the next two weeks, so PR turnaround won't be great during that time.

**Comment by cardoso-neto** [comment]
Got it. I'll keep you posted.

**Comment by rgommers** [comment]
> @rgommers Did it work with the Numba C code as well?
> I've come across a few other tools that also promise to do that, but I cannot speak to their effectiveness.
> [github/autoflake/autostar.py](https://github.com/fake-name/autoflake/blob/master/autostar.py)
> [github/quentinsf/dewildcard](https://github.com/quentinsf/dewildcard)

Not sure. There's not much C code in Numba. @asmeurer can you comment?

> > modules do from m import f and accidentally expose f in their context
> 
> @ezyang Well, great. This is bad and we would avoid it by using explicit imports.
> 
> We could also enable a check on the pull requests so it'll flag bad star imports.

Another, perhaps more comprensive, way to check both that nothing leaks into a namespace and that there are docs for everything that's meant to be public is something like https://github.com/numpy/numpy/blob/master/tools/refguide_check.py. That checks `__all__` dicts for completeness again `dir(public_module)`, and then compares against doc entries. The way to grab doc entries (`find_names` in there) would need changing - probably to looking at the Sphinx inventory file - but it's a useful set of checks that could be adapted.

**Comment by asmeurer** [comment]
There's an open issue about making it work with C extensions https://github.com/asmeurer/removestar/issues/9. I think it might be straightforward to make it work if you can do an in-place build.

**Comment by cardoso-neto** [comment]
I just tried `removestar` on `torch/quantization/qconfig.py`
```python 
from .observer import *
from .fake_quantize import *
```
Became:
```python
from .observer import (PlaceholderObserver, default_debug_observer, default_dynamic_quant_observer,
                       default_float_qparams_observer, default_observer,
                       default_per_channel_weight_observer, default_placeholder_observer,
                       default_weight_observer)
from .fake_quantize import (FakeQuantize, HistogramObserver, MovingAverageMinMaxObserver,
                            default_fake_quant, default_per_channel_weight_fake_quant,
                            default_weight_fake_quant, torch)
```
Which is not properly indented, but that's ok. What was weird to me was its grouping of classes and functions, instead of using a purely alphabetical sorting. Should we go for its grouped import or just alphabetical?

Aesthetics aside, it threw some worrying warnings:
```
Warning: torch/quantization/qconfig.py: 'MovingAverageMinMaxObserver' comes from multiple modules: '.observer', '.fake_quantize'. Using '.fake_quantize'.
Warning: torch/quantization/qconfig.py: 'HistogramObserver' comes from multiple modules: '.observer', '.fake_quantize'. Using '.fake_quantize'.
Warning: torch/quantization/qconfig.py: 'torch' comes from multiple modules: '.observer', '.fake_quantize'. Using '.fake_quantize'.
```
Its confusion is understandable, and a human (me) can easily discern from which module is best to import what.
It also obviously doesn't work with `__init__.py` files, but that's ok; I think I can easily do those manually as well.

**Comment by rgommers** [comment]
> What was weird to me was its grouping of classes and functions, instead of using a purely alphabetical sorting. Should we go for its grouped imports or for alphabetical only?

I like the current output (separate classes and functions), more readable. Try `dir(torch)` in an interpreter, it will behave the same.

**Comment by asmeurer** [comment]
It's just doing a `sort()` on the names which is case sensitive. You could also try running the import lines through something like `black` or `isort`.

**Comment by cardoso-neto** [comment]
I just did the `torch/cuda` module. I added the commit on my fork's [replace-star-imports](https://github.com/pytorch/pytorch/compare/master...cardoso-neto:replace-star-imports) branch.
I took the liberty of removing the `__all__` override, as it wasn't really doing anything anymore.

**Comment by rgommers** [comment]
> I also removed the `__all__` override, as it wasn't really doing anything anymore.

Having `__all__` is still useful, even if only to document which functionality in a given file is meant to be public. If it's already there, I'd leave it as is.

**Comment by cardoso-neto** [comment]
Got it. I edited the commit. It's back in there already.
